File: starter_kits/Python3/hlt/positionals.py
from . import commands


class Direction:
    """
    Holds positional tuples in relation to cardinal directions
    """
    North = (0, -1)
    South = (0, 1)
    East = (1, 0)
    West = (-1, 0)

    Still = (0, 0)

    @staticmethod
    def get_all_cardinals():
        """
        Returns all contained items in each cardinal
        :return: An array of cardinals
        """
        return [Direction.North, Direction.South, Direction.East, Direction.West]

    @staticmethod
    def convert(direction):
        """
        Converts from this direction tuple notation to the engine's string notation
        :param direction: the direction in this notation
        :return: The character equivalent for the game engine
        """
        if direction == Direction.North:
            return commands.NORTH
        if direction == Direction.South:
            return commands.SOUTH
        if direction == Direction.East:
            return commands.EAST
        if direction == Direction.West:
            return commands.WEST
        if direction == Direction.Still:
            return commands.STAY_STILL
        else:
            raise IndexError

    @staticmethod
    def invert(direction):
        """
        Returns the opposite cardinal direction given a direction
        :param direction: The input direction
        :return: The opposite direction
        """
        if direction == Direction.North:
            return Direction.South
        if direction == Direction.South:
            return Direction.North
        if direction == Direction.East:
            return Direction.West
        if direction == Direction.West:
            return Direction.East
        if direction == Direction.Still:
            return Direction.Still
        else:
            raise IndexError


class Position:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def directional_offset(self, direction):
        """
        Returns the position considering a Direction cardinal tuple
        :param direction: the direction cardinal tuple
        :return: a new position moved in that direction
        """
        return self + Position(*direction)

    def get_surrounding_cardinals(self):
        """
        :return: Returns a list of all positions around this specific position in each cardinal direction
        """
        return [self.directional_offset(current_direction) for current_direction in Direction.get_all_cardinals()]

    def __add__(self, other):
        return Position(self.x + other.x, self.y + other.y)

    def __sub__(self, other):
        return Position(self.x - other.x, self.y - other.y)

    def __iadd__(self, other):
        self.x += other.x
        self.y += other.y
        return self

    def __isub__(self, other):
        self.x -= other.x
        self.y -= other.y
        return self

    def __abs__(self):
        return Position(abs(self.x), abs(self.y))

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

    def __ne__(self, other):
        return not self.__eq__(other)

    def __repr__(self):
        return "{}({}, {})".format(self.__class__.__name__,
                                   self.x,
                                   self.y)

File: starter_kits/Ruby/hlt/positionals.rb
require_relative 'commands'
module Positionals
  # Holds positional tuples in relation to cardinal directions
  class Direction
    NORTH = [ 0, -1]
    SOUTH = [ 0,  1]
    EAST  = [ 1,  0]
    WEST  = [-1,  0]

    STILL = [ 0,  0]

    # Returns all contained items in each cardinal
    # @return [Array] An array of all cardinals
    def self.get_all_cardinals
      [
        Direction::NORTH,
        Direction::SOUTH,
        Direction::EAST,
        Direction::WEST
      ]
    end

    # Converts from this direction tuple notation to the
    # engine's string notation.
    # @param direction: The direction in this notation.
    # @return [String] The character equivalent for the game engine.
    def self.convert(direction)
      case direction
      when Direction::NORTH
        Commands::NORTH
      when Direction::SOUTH
        Commands::SOUTH
      when Direction::EAST
        Commands::EAST
      when Direction::WEST
        Commands::WEST
      when Direction::STILL
        Commands::STAY_STILL
      else
        raise "Invalid direction tuple"
      end
    end

    # Returns the opposite cardinal direction from a given direction.
    # @param direction: The input direction
    # @return [Direction] The opposite direction.
    def self.invert
      case direction
      when Direction::NORTH
        Direction::SOUTH
      when Direction::SOUTH
        Direction::NORTH
      when Direction::EAST
        Direction::WEST
      when Direction::WEST
        Direction::EAST
      when Direction::STILL
        Direction::STILL
      else
        raise "Invalid direction"
      end
    end
  end

  class Position
    attr_reader :x, :y

    def initialize(x, y)
      @x = x
      @y = y
    end

    # Returns the position considering a Direction cardinal tuple
    # @param direction: The direction cardinal tuple
    # @return [Position] A new position moved in that direction
    def directional_offset(direction)
      self + new(*direction)
    end

    # @return Returns a list of all positions around this specific
    # position in each cardinal direction.
    def get_surrounding_cardinals
      Direction.get_all_cardinals.map do |direction|
        directional_offset(direction)
      end
    end

    def +(other)
      new(@x + other.x, @y + other.y)
    end

    def -(other)
      new(@x - other.x, @y - other.y)
    end

    def abs
      new(@x.abs, @y.abs)
    end

    def ==(other)
      @x == other.x && @y == other.y
    end

    def !=(other)
      !self.==(other)
    end
  end
end
